tests/test_instrument.py

```diff
@@ -178,6 +188,97 @@ def test_collection_timestamps(self):
         with self.assertRaises(TypeError):
             Instrument.formatCollectionTimestamp(0)
 
+    def test_dimension_packer_config_defaults(self):
+        """Test the ObservationDimensionPacker class and the Instrument-based
+        systems for constructing it, in the case where the Instrument-defined
+        default is used.
+        """
+        registry_config = RegistryConfig()
+        registry_config["db"] = "sqlite://"
+        registry = Registry.createFromConfig(registry_config)
+        self.instrument.register(registry)
+        config = DimensionPackerTestConfig()
+        instrument_data_id = registry.expandDataId(instrument=self.name)
+        record = instrument_data_id.records["instrument"]
+        self.check_dimension_packers(
+            registry,
+            # Test just one packer-construction signature here as that saves us
+            # from having to insert dimension records for visits, exposures,
+            # and detectors for this test.
+            visit_packers=[config.packer.apply(instrument_data_id, is_exposure=False)],
+            exposure_packers=[config.packer.apply(instrument_data_id, is_exposure=True)],
+            n_detectors=record.detector_max,
+            n_visits=record.visit_max,
+            n_exposures=record.exposure_max,
+        )
+
+    def test_dimension_packer_config_override(self):
+        """Test the ObservationDimensionPacker class and the Instrument-based
+        systems for constructing it, in the case where configuration overrides
+        the Instrument's default.
+        """
+        registry_config = RegistryConfig()
+        registry_config["db"] = "sqlite://"
+        registry = Registry.createFromConfig(registry_config)
+        # Intentionally do not register instrument or insert any other
+        # dimension records to ensure we don't need them in this mode.
+        config = DimensionPackerTestConfig()
+        config.packer["observation"].n_observations = 16
+        config.packer["observation"].n_detectors = 4
+        config.packer.name = "observation"
+        instrument_data_id = DataCoordinate.standardize(instrument=self.name, universe=registry.dimensions)
+        full_data_id = DataCoordinate.standardize(instrument_data_id, detector=1, visit=3, exposure=7)
+        visit_data_id = full_data_id.subset(full_data_id.universe.extract(["visit", "detector"]))
+        exposure_data_id = full_data_id.subset(full_data_id.universe.extract(["exposure", "detector"]))
+        self.check_dimension_packers(
+            registry,
+            visit_packers=[
+                config.packer.apply(visit_data_id),
+                config.packer.apply(full_data_id),
+                config.packer.apply(instrument_data_id, is_exposure=False),
```

Comment:
I must be misunderstanding something here, because I don't see how this works; there's no visit in the dataId to pack.

Reply:
Calling `apply` here just constructs the `ObservationDimensionPacker` instance.  It isn't until `pack` is called that you need the full data ID.

I'm guessing this is confusing in part because you've just looked at the `meas_base` PR, where calling `apply` on a config field ends up both constructing a packer and calling `pack` on it.  It all boils down to `DimensionPacker` being a per-instrument-and-dimensions or (or per-skymap-and-dimensions) thing, while `IdGenerator` is a per-data-ID thing.

Reply:
Might be worth a comment to that effect here? Or maybe I'm just confused by having reviewed all this.